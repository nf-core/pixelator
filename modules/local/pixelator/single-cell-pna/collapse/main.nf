process PIXELATOR_PNA_COLLAPSE {
    tag "${meta.id}"
    label 'process_medium'

    containerOptions 'shm-size': 2.GB

    // TODO: Add conda
    // conda "bioconda::pixelator=0.18.2"
    container "${workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container
        ? 'quay.io/pixelgen-technologies/pixelator:0.21.2'
        : 'quay.io/pixelgen-technologies/pixelator:0.21.2'}"

    input:
    tuple val(meta), path(reads), path(panel_file), val(panel), val(design)

    output:
    tuple val(meta), path("collapse/*.parquet", arity: '1..*'),     emit: collapsed
    tuple val(meta), path("collapse/*.report.json", arity: '1..*'), emit: report_json
    tuple val(meta), path("collapse/*.meta.json"),                  emit: metadata_json
    tuple val(meta), path("*pixelator-collapse.log"),               emit: log

    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    assert meta.design != null

    prefix = task.ext.prefix ?: "${meta.id}"
    def args = task.ext.args ?: ''
    def read_args = reads.join(' ')
    def panel_opt = (panel
        ? "--panel ${panel}"
        : panel_file
            ? "--panel ${panel_file}"
            : "")

    """
    pixelator \\
        --log-file ${prefix}.pixelator-collapse.log \\
        --verbose \\
        single-cell-pna \\
        collapse \\
        --threads ${task.cpus} \\
        --output . \\
        --design ${design} \\
        ${panel_opt} \\
        ${args} \\
        ${read_args}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        pixelator: \$(echo \$(pixelator --version 2>/dev/null) | sed 's/pixelator, version //g' )
    END_VERSIONS
    """

    stub:
    prefix = "${reads.name.replace('.parquet', '')}"

    """
    mkdir "collapse"

    touch collapse/${prefix}.report.json
    touch collapse/${prefix}.meta.json
    touch collapse/${prefix}.parquet
    touch ${prefix}.pixelator-collapse.log

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        pixelator: \$(echo \$(pixelator --version 2>/dev/null) | sed 's/pixelator, version //g' )
    END_VERSIONS
    """
}
