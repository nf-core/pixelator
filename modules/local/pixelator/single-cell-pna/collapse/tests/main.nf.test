nextflow_process {

    name "Test Process PIXELATOR_PNA_COLLAPSE"
    script "../main.nf"
    process "PIXELATOR_PNA_COLLAPSE"
    tag "modules"
    tag "pixelator"
    tag "pixelator/pna"
    tag "pixelator/single_cell_pna_collapse"

    test("PNA collapse - small test - single file") {
        when {
            process {
                """
                input[0] = [
                    [ id:'PNA055_Sample07_filtered_S7', design:'pna-2', panel:'proxiome-immuno-155', technology:'pna' ],
                    [
                        file(params.pipelines_testdata_base_path + 'testdata/pna/modules/demux/PNA055_Sample07_filtered_S7.demux.m1.part_000.parquet', checkIfExists: true)
                    ],
                    [],                     // panel_file
                    'proxiome-immuno-155',   // panel
                    'pna-2'                 // design
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                {
                    with(process.out.collapsed) {
                        assert get(0).get(1).every { file(it).exists() }
                    }
                },
                { assert snapshot(process.out.metadata_json).match("metadata") },
                { assert snapshot(process.out.versions).match("versions") },
                {
                    with(process.out.log) {
                        assert path(get(0).get(1)).readLines().last().contains("Finished pixelator collapse")
                    }
                },
                {
                    with(process.out.report_json) {
                        assert snapshot(path(get(0).get(1).get(0)).readLines()[0..2]).match("header")
                    }
                }
            )
        }
    }

    test("PNA collapse - small test - multiple files") {
        when {
            process {
                """
                input[0] = [
                    [ id:'PNA055_Sample07_filtered_S7', design:'pna-2', panel:'proxiome-immuno-155', technology:'pna' ],
                    [
                        file(params.pipelines_testdata_base_path + 'testdata/pna/modules/demux/PNA055_Sample07_filtered_S7.demux.m1.part_000.parquet', checkIfExists: true),
                        file(params.pipelines_testdata_base_path + 'testdata/pna/modules/demux/PNA055_Sample07_filtered_S7.demux.m1.part_001.parquet', checkIfExists: true)
                    ],
                    [],                      // panel_file
                    'proxiome-immuno-155',   // panel
                    'pna-2'                  // design
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                {
                    with(process.out.collapsed) {
                        assert get(0).get(1).every { file(it).exists() }
                    }
                },
                { assert snapshot(process.out.metadata_json).match("multi_metadata") },
                { assert snapshot(process.out.versions).match("multi_versions") },
                {
                    with(process.out.log) {
                        assert path(get(0).get(1)).readLines().last().contains("Finished pixelator collapse")
                    }
                },
                {
                    with(process.out.report_json) {
                        assert snapshot(path(get(0).get(1).get(0)).readLines()[0..2]).match("multi_header")
                    }
                }
            )
        }
    }
}
