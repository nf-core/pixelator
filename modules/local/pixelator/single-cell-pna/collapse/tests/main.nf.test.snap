{
    "multi_metadata": {
        "content": [
            [
                [
                    {
                        "id": "PNA055_Sample07_filtered_S7",
                        "design": "pna-2",
                        "panel": "proxiome-immuno-155",
                        "technology": "pna"
                    },
                    [
                        "PNA055_Sample07_filtered_S7.collapse.m1.part_000.meta.json:md5,ef6d3db8cfb46d551bc6fb160f93d44a",
                        "PNA055_Sample07_filtered_S7.collapse.m1.part_001.meta.json:md5,ef6d3db8cfb46d551bc6fb160f93d44a"
                    ]
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-20T11:11:51.603558511"
    },
    "metadata": {
        "content": [
            [
                [
                    {
                        "id": "PNA055_Sample07_filtered_S7",
                        "design": "pna-2",
                        "panel": "proxiome-immuno-155",
                        "technology": "pna"
                    },
                    "PNA055_Sample07_filtered_S7.collapse.m1.part_000.meta.json:md5,ef6d3db8cfb46d551bc6fb160f93d44a"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-20T09:39:49.273740685"
    },
    "multi_header": {
        "content": [
            [
                "{",
                "    \"sample_id\": \"PNA055_Sample07_filtered_S7\",",
                "    \"product_id\": \"single-cell-pna\","
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-19T15:58:03.194867001"
    },
    "versions": {
        "content": [
            [
                "versions.yml:md5,c3ca37ec135b9455a855ff59c6ee8019"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-06-26T11:59:00.136057252"
    },
    "multi_versions": {
        "content": [
            [
                "versions.yml:md5,c3ca37ec135b9455a855ff59c6ee8019"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-06-26T11:59:21.486766006"
    },
    "header": {
        "content": [
            [
                "{",
                "    \"sample_id\": \"PNA055_Sample07_filtered_S7\",",
                "    \"product_id\": \"single-cell-pna\","
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-20T09:39:49.312268752"
    }
}